network/msgCompressor.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package network
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/DataDog/zstd"
+
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/protocol"
+)
+
+var zstdCompressionMagic = [4]byte{0x28, 0xb5, 0x2f, 0xfd}
+
+const zstdCompressionLevel = zstd.BestSpeed
+
+// checkCanCompress checks if there is an proposal payload message and peers supporting compression
+func checkCanCompress(request broadcastRequest, peers []*wsPeer) bool {
+	canCompress := false
+	hasPP := false
+	for _, tag := range request.tags {
+		if tag == protocol.ProposalPayloadTag {
+			hasPP = true
+			break
+		}
+	}
+	// if have proposal payload check if there are any peers supporting compression
+	if hasPP {
+		for _, peer := range peers {
+			if peer.pfProposalCompressionSupported() {
+				canCompress = true
+				break
+			}
+		}
+	}
+	return canCompress
+}
+
+// zstdCompressMsg returns a concatenation of a tag and compressed data
+func zstdCompressMsg(tbytes []byte, d []byte) ([]byte, string) {
+	bound := zstd.CompressBound(len(d))
+	if bound < len(d) {
+		// although CompressBound allocated more than the src size, this is an implementation detail.
+		// increase the buffer size to always have enough space for the raw data if compression fails.
+		bound = len(d)
+	}
+	mbytesComp := make([]byte, len(tbytes)+bound)
+	copy(mbytesComp, tbytes)
+	comp, err := zstd.CompressLevel(mbytesComp[len(tbytes):], d, zstdCompressionLevel)
+	if err != nil {
+		// fallback and reuse non-compressed original data
+		logMsg := fmt.Sprintf("failed to compress into buffer of len %d: %v", len(d), err)
+		copied := copy(mbytesComp[len(tbytes):], d)
+		return mbytesComp[:len(tbytes)+copied], logMsg
+	}
+	mbytesComp = mbytesComp[:len(tbytes)+len(comp)]
+	return mbytesComp, ""
+}
+
+// MaxDecompressedMessageSize defines a maximum decompressed data size
+// to prevent zip bombs
+const MaxDecompressedMessageSize = 20 * 1024 * 1024 // some large enough value
```

Comment:
I was going to suggest using some amount more than the largest proto.MaxTxnBytesPerBlock, because I had expected we already calculated that somewhere, for an allocbound.  But we don't.

So instead, perhaps just put "consider MaxTxnBytesPerBlock" in the comment, so we find it when grepping if we consider changing it.

Reply:
done in https://github.com/algorand/go-algorand/pull/4650